escodegen.js

```diff
@@ -1836,8 +1838,34 @@
         BinaryExpression: function (expr, precedence, flags) {
             var result, leftPrecedence, rightPrecedence, currentPrecedence, fragment, leftSource;
             currentPrecedence = BinaryPrecedence[expr.operator];
-            leftPrecedence = expr.operator === '**' ? Precedence.Postfix : currentPrecedence;
-            rightPrecedence = expr.operator === '**' ? currentPrecedence : currentPrecedence + 1;
+
+            leftPrecedence = currentPrecedence;
+            rightPrecedence = currentPrecedence + 1;
+
+            switch (expr.operator) {
+                case '**':
+                    leftPrecedence = Precedence.Postfix;
+                    rightPrecedence = currentPrecedence;
+                    break;
+
+                case '??':
+                    if (expr.left.operator === '||' || expr.left.operator === '&&') {
+                        leftPrecedence = BinaryPrecedence[expr.left.operator] + 1;
+                    }
+
+                    if (expr.right.operator === '||' || expr.right.operator === '&&') {
```

Comment:
You don't need to include `expr.right.operator === '||'` here for the same reason you didn't need to include `expr.right.operator === '??'` below.

Reply:
https://github.com/estools/escodegen/pull/417/commits/5646f113ae78eca9046da1b8e4ea20d4baec87ad